Improved quantize script

SuajCarrot opened the pull request:
I improved the quantize script by adding error handling and allowing to select many models for quantization at once in the command line. I also converted it to Python for generalization as well as extensibility.

SuajCarrot committed:
Improved quantize script

I improved the quantize script by adding error handling and allowing to select many models for quantization at once in the command line. I also converted it to Python for generalization as well as extensibility.

mattsta commented on quantize.py:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/python3
```
should be `#!/usr/bin/env python3`

SuajCarrot replied:
Corrected.

mattsta commented on quantize.py:
```diff
@@ -0,0 +1,59 @@
+        description='This script quantizes a model.'
+    )
+
+    parser.add_argument("models", nargs='+', dest='models')
```
could use choices here (if we want to limit to the 7/13/30/65 models): https://docs.python.org/3/library/argparse.html#choices

something like:

```python
parser.add_argument("models", nargs='+', dest='models', choices=("7B", "13B", "30B", "65B"))
```

SuajCarrot replied:
Added.

mattsta commented on quantize.py:
```diff
@@ -0,0 +1,59 @@
+
+        for i in os.listdir(model_path):
+            subprocess.run(
+                ["./quantize", i, i.replace("f16", "q4_0"), "2"],
```
also the `quantize` script isn't necessarily in the current directory (example: CMake builds).

would be useful if it is a parameter (with a default at `./`) so we could point it elsewhere easily too.

SuajCarrot replied:
Functionality added; you also gave me the idea of adding an argument that lets the user select the directory where the models are located, which defaults to `os.path.join(os.getcwd(), "models")`!

mattsta commented on quantize.py:
```diff
@@ -0,0 +1,59 @@
+
+    for model in args.models:
+
+        model_path = os.path.join("models", model, "ggml-model-f16.bin")
```
looks like `model_path` is a file, but then you're doing `os.listdir()` on a file? not sure how that would work.

also we would want to limit the directory listing to just a glob of the proper matching filenames since there's many other things in these directories.

SuajCarrot replied:
I think I missunderstood the original script (`for i in `ls models/$1/ggml-model-f16.bin*`), looks like it is in fact a file. Thank you for the observation.

mattsta commented on quantize.py:
```diff
@@ -0,0 +1,59 @@
+
+        model_path = os.path.join("models", model, "ggml-model-f16.bin")
+
+        for i in os.listdir(model_path):
```
there's another PR about parallelizing the quantizations. here it would be easy just to wrap this in a `multiprocessing.Pool()` (with the subprocess calls extracted out to a top-level function): https://docs.python.org/3/library/multiprocessing.html

SuajCarrot replied:
I don't understand exactly what can be parallelized here, is it quantizing many models at the same time? The code from the latest commit removed the loop that acts upon the incorrect listing of the file (`for i in os.listdir(model_path)`) by the way.

SuajCarrot committed:
Fixes and improvements based on Matt's observations

Fixed and improved many things in the script based on the reviews made by @mattsta. The parallelization suggestion is still to be revised, but code for it was still added (commented).

SuajCarrot committed:
Small fixes to the previous commit

SuajCarrot committed:
Corrected to use the original glob pattern

The original Bash script uses a glob pattern to match files that have endings such as ...bin.0, ...bin.1, etc. That has been translated correctly to Python now.

sw commented:
I think it's a good idea to remove the requirement for a Unix shell.
See also #285 which would be made obsolete by this, assuming the Python script works on Windows?

I suggest to remove quantize.sh and update the readme to use the python script.

SuajCarrot committed:
Added support for Windows and updated README to use this script

New code to set the name of the quantize script binary depending on the platform has been added (quantize.exe if working on Windows) and the README.md file has been updated to use this script instead of the Bash one.

SuajCarrot commented:
Thank you for your comment @sw, I just pushed a commit that applies the changes you suggested. Let me know if there's anything else that should be done to achieve full compatibility with Windows.

sw commented on quantize.py:
```diff
@@ -0,0 +1,126 @@
+    f16 model to quantize.
+    """
+
+    new_quantized_model_path = f16_model_part_path.replace("16", "q4_0")
```
This should be `.replace("f16", "q4_0")`, or there will be an extra 'f' in the output filename

SuajCarrot replied:
Fixed.

sw commented:
.devops/tools.sh needs to be updated as well, that calls quantize.sh. Can you update your branch and do that as well please?

SuajCarrot committed:
Merge remote-tracking branch 'upstream/master' into suajcarrot-changes

sw commented on quantize.py:
```diff
@@ -0,0 +1,126 @@
+    new_quantized_model_path = f16_model_part_path.replace("16", "q4_0")
+    subprocess.run(
+        [script_path, f16_model_part_path, new_quantized_model_path, "2"],
+        shell=True, check=True
```
`shell=True` doesn't work for me (on Linux), because you would have to give the entire command line as a string instead of a list of strings.

SuajCarrot replied:
Removed, you're right.

SuajCarrot committed:
Fixed a typo and removed shell=True in the subprocess.run call

Fixed a typo regarding the new filenames of the quantized models and removed the shell=True parameter in the subprocess.run call as it was conflicting with the list of parameters.

SuajCarrot committed:
Corrected previous commit

SuajCarrot committed:
Small tweak: changed the name of the program in argparse

This was making the automatic help message to be suggesting the program's usage as being literally "$ Quantization Script [arguments]". It should now be something like "$ python3 quantize.py [arguments]".

ggerganov commented:
Should we merge now or wait for someone to test on Windows?

@SuajCarrot maybe keep the `.sh` for now and add a comment that it is deprecated. We will remove it later.

SuajCarrot commented:
Thank you for merging! Should I create another pull request with the Bash script added back as well as its deprecation notice in the README?

ggerganov commented:
It was confirmed in #285  that it works on Windows, so no need to do it

SuajCarrot commented:
That's great, thank you.
